src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -140,16 +230,826 @@ Pipe ReadFromMergeTree::read()
     return pipe;
 }
 
+Pipe ReadFromMergeTree::spreadMarkRangesAmongStreams(
+    RangesInDataParts && parts_with_ranges,
+    const Names & column_names)
+{
+    const auto & settings = context->getSettingsRef();
+
+    /// Count marks for each part.
+    std::vector<size_t> sum_marks_in_parts(parts_with_ranges.size());
+    size_t sum_marks = 0;
+    size_t total_rows = 0;
+
+    const auto data_settings = data.getSettings();
+    size_t adaptive_parts = 0;
+    for (size_t i = 0; i < parts_with_ranges.size(); ++i)
+    {
+        total_rows += parts_with_ranges[i].getRowsCount();
+        sum_marks_in_parts[i] = parts_with_ranges[i].getMarksCount();
+        sum_marks += sum_marks_in_parts[i];
+
+        if (parts_with_ranges[i].data_part->index_granularity_info.is_adaptive)
+            ++adaptive_parts;
+    }
+
+    size_t index_granularity_bytes = 0;
+    if (adaptive_parts > parts_with_ranges.size() / 2)
+        index_granularity_bytes = data_settings->index_granularity_bytes;
+
+    const size_t max_marks_to_use_cache = MergeTreeDataSelectExecutor::roundRowsOrBytesToMarks(
+        settings.merge_tree_max_rows_to_use_cache,
+        settings.merge_tree_max_bytes_to_use_cache,
+        data_settings->index_granularity,
+        index_granularity_bytes);
+
+    const size_t min_marks_for_concurrent_read = MergeTreeDataSelectExecutor::minMarksForConcurrentRead(
+        settings.merge_tree_min_rows_for_concurrent_read,
+        settings.merge_tree_min_bytes_for_concurrent_read,
+        data_settings->index_granularity,
+        index_granularity_bytes,
+        sum_marks);
+
+    bool use_uncompressed_cache = settings.use_uncompressed_cache;
+    if (sum_marks > max_marks_to_use_cache)
+        use_uncompressed_cache = false;
```

Comment:
This quite huge part of code looks identical in 3 places. Is it possible to extract it to separate function?

Reply:
ok, I will try